app/src/main/java/com/example/composetemplate/managers/TokenManager.kt

```diff
@@ -0,0 +1,36 @@
+package com.example.composetemplate.managers
+
+interface TokenFetcher {
+    suspend fun fetchToken(): TokenData
+}
+
+data class TokenData(
+    val token: String?,
+    val expirationTimeLong: Long?
+)
+
+class TokenManager(private val tokenFetcher: TokenFetcher) {
+    private var token: String? = null
+    private var expirationTime: Long? = 0
+
+    suspend fun getToken(): String {
+        if (isTokenExpired()) {
+            val (newToken, newExpirationTime) = tokenFetcher.fetchToken()
+            if (newToken != null) {
+                saveToken(newToken, newExpirationTime)
+            } else {
+                throw Exception("Failed to fetch a valid token")
```

Comment:
I know we discussed this, but I’m writing it down to keep it in mind. You should move the strings to the constants file and consider creating a specific exception for token-related cases.